From f5648ce7f4f27e55124f8229f57e849443589366 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:07:59 -0400
Subject: [powerpc] Fix endian issues in power7/8 machine check handler

Message-id: <1410545655-205645-251-git-send-email-dzickus@redhat.com>
Patchwork-id: 94279
O-Subject: [RHEL7 PATCH 250/626] powerpc: Fix endian issues in power7/8 machine check handler
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit a68c33f3592eef63304a5f5ab68466539ccac56c
Author: Anton Blanchard <anton@samba.org>
Date:   Mon Dec 16 10:47:54 2013 +1100

    powerpc: Fix endian issues in power7/8 machine check handler

    The SLB save area is shared with the hypervisor and is defined
    as big endian, so we need to byte swap on little endian builds.

    Signed-off-by: Anton Blanchard <anton@samba.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/mce_power.c b/arch/powerpc/kernel/mce_power.c
index b36e777..27c93f4 100644
--- a/arch/powerpc/kernel/mce_power.c
+++ b/arch/powerpc/kernel/mce_power.c
@@ -50,12 +50,12 @@ static void flush_and_reload_slb(void)
  if (!slb)
   return;
 
- n = min_t(u32, slb->persistent, SLB_MIN_SIZE);
+ n = min_t(u32, be32_to_cpu(slb->persistent), SLB_MIN_SIZE);
 
  /* Load up the SLB entries from shadow SLB */
  for (i = 0; i < n; i++) {
-  unsigned long rb = slb->save_area[i].esid;
-  unsigned long rs = slb->save_area[i].vsid;
+  unsigned long rb = be64_to_cpu(slb->save_area[i].esid);
+  unsigned long rs = be64_to_cpu(slb->save_area[i].vsid);
 
   rb = (rb & ~0xFFFul) | i;
   asm volatile("slbmte %0,%1" : : "r" (rs), "r" (rb));
-- 
1.7.1